mr/src/main/java/org/elasticsearch/hadoop/rest/bulk/BulkProcessor.java

```diff
@@ -210,6 +212,16 @@ public BulkResponse tryFlush() {
                     // Log messages, and if wait time is set, perform the thread sleep.
                     initFlushOperation(bulkLoggingID, retryOperation, retries.size(), waitTime);
 
+                    // Check if should wait before sending the request.
+                    try {
+                        if (this.batchWriteWaitTime != 0 && !retryOperation)
+                            Thread.sleep(batchWriteWaitTime);
+                    }
```

Comment:
I think in this case we would want something more robust to measure the time between bulk operations. If the wait time is 30 seconds but it takes 20 seconds to generate enough records to perform a bulk operation, then the total time between bulks is 50 seconds instead of 30. Since the time spent reaching the batch size limit is variable, the time to wait between each bulk operation is variable as well.

Additionally, the bulk operation does not have a consistent size in bytes or number - rather, the bulk operation is executed based on which ever limit is hit first. This variability of batch size affects how long it takes for ES to persist the data. When a batch operation would be executed and with how many records is completely disconnected from the wait time, which means that this property can become a moving target in terms of getting the right value.

I think instead of that, we would want to measure and cap the rate of ingestion to a configured "event per time unit" target rate. This is similar to how Elasticsearch's Reindex operation does throttling (https://www.elastic.co/guide/en/elasticsearch/reference/7.5/docs-reindex.html#docs-reindex-api-query-params). This is much easier to reason about and does a more consistent job of cohesively managing the rate of indexing than just a wait time.

Reply:
I agree it is not good solution to wait between bulks, but i don`t know if we can implement "event per time unit" because it is a little bit fake ( i mean we have x Executors that each executor will write in that index rate).
How about starting a timer after each bulk finish and before the next bulk writing wait till timer finish?